addRetNodesToCurrentCFG is broken

Hi @lucaneg,

I discovere a bug inside the function addRetNodesToCurrentCFG.

If in a function we have a return inside a conditional branch, for example:

```
def test():
    if True:
        return
    a = 10

test()
```
The analysis will throw an error:
```
~/Projects/pylisa/pylisa (master ✗) ./gradlew run --args="ros_examples/test_ret.py test_ret_results"

> Task :run FAILED
2023-09-21 15:35:38,209 [ INFO] PyToCFG setup... 
2023-09-21 15:35:38,362 [ INFO] Reading file... ros_examples/test_ret.py 
line 6:6 no viable alternative at input '<EOF>'
2023-09-21 15:35:38,415 [ INFO] Parsing stmt lists... [start] 
2023-09-21 15:35:38,419 [ INFO] Parsing stmt lists...: 1/2 
2023-09-21 15:35:38,432 [ INFO] Parsing stmt lists...: 2/2 
2023-09-21 15:35:38,436 [ INFO] Parsing stmt lists... [stop] [2 Global stmt in 21ms 539972ns] 
2023-09-21 15:35:38,466 [ INFO] LiSA configuration:
  syntacticChecks (0)
  semanticChecks (2):
    DataframeDumper
    DataframeStructureConstructor
  analysisGraphs: NONE
  serializeInputs: false
  serializeResults: false
  jsonOutput: true
  workdir: test_ret_results
  wideningThreshold: 5
  recursionWideningThreshold: 5
  glbThreshold: 5
  descendingPhaseType: NONE
  fixpointWorkingSet: DuplicateFreeFIFOWorkingSet
  openCallPolicy: ReturnTopPolicy
  optimize: true
  hotspots: unset
  dumpForcesUnwinding: false 
2023-09-21 15:35:38,498 [ INFO] Analysis time [start] 
2023-09-21 15:35:38,499 [ INFO] Finalizing input program [start] 
Exception in thread "main" it.unive.lisa.AnalysisException: LiSA has encountered an exception while executing the analysis
        at it.unive.lisa.LiSA.run(LiSA.java:101)
        at it.unive.pylisa.PyLiSA.main(PyLiSA.java:54)
Caused by: it.unive.lisa.AnalysisExecutionException: Unable to finalize target program
        at it.unive.lisa.LiSARunner.lambda$finalizeApp$10(LiSARunner.java:280)
        at it.unive.lisa.logging.TimerLogger.execAux(TimerLogger.java:116)
        at it.unive.lisa.logging.TimerLogger.execAction(TimerLogger.java:57)
        at it.unive.lisa.logging.TimerLogger.execAction(TimerLogger.java:29)
        at it.unive.lisa.LiSARunner.finalizeApp(LiSARunner.java:276)
        at it.unive.lisa.LiSARunner.run(LiSARunner.java:104)
        at it.unive.lisa.LiSA.lambda$run$0(LiSA.java:99)
        at it.unive.lisa.logging.TimerLogger.lambda$execSupplier$0(TimerLogger.java:108)
        at it.unive.lisa.logging.TimerLogger.execAux(TimerLogger.java:116)
        at it.unive.lisa.logging.TimerLogger.execSupplier(TimerLogger.java:108)
        at it.unive.lisa.logging.TimerLogger.execSupplier(TimerLogger.java:73)
        at it.unive.lisa.LiSA.run(LiSA.java:99)
        ... 1 more
Caused by: it.unive.lisa.program.ProgramValidationException: untyped ros_examples/test_ret::test() [at 'ros_examples/test_ret.py':1:0] contains a node with no followers that is not execution-stopping: a = 10
        at it.unive.lisa.program.cfg.CFG.validate(CFG.java:659)
        at it.unive.lisa.program.language.validation.BaseValidationLogic.validate(BaseValidationLogic.java:406)
        at it.unive.lisa.program.language.validation.BaseValidationLogic.validateAndFinalize(BaseValidationLogic.java:179)
        at it.unive.lisa.program.language.validation.BaseValidationLogic.validateAndFinalize(BaseValidationLogic.java:163)
        at it.unive.lisa.LiSARunner.lambda$finalizeApp$10(LiSARunner.java:278)
        ... 12 more

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':run'.
> Process 'command '/usr/local/Cellar/openjdk@17/17.0.8/libexec/openjdk.jdk/Contents/Home/bin/java'' finished with non-zero exit value 1

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 1s
4 actionable tasks: 1 executed, 3 up-to-date
~/Projects/pylisa/pylisa (master ✗) 
```

This because *currentCFG.getAllExitpoints()* is not empty, and the code will not add the Return statement at the end of the function:
```
private void addRetNodesToCurrentCFG() {
		if (currentCFG.getAllExitpoints().isEmpty()) {
			Ret ret = new Ret(currentCFG, currentCFG.getDescriptor().getLocation());
			if (currentCFG.getNodesCount() == 0) {
				// empty method, so the ret is also the entrypoint
				currentCFG.addNode(ret, true);
			} else {
				// every non-throwing instruction that does not have a follower
				// is ending the method
				Collection<Statement> preExits = new LinkedList<>();
				for (Statement st : currentCFG.getNodes())
					if (!st.stopsExecution() && currentCFG.followersOf(st).isEmpty())
						preExits.add(st);
				currentCFG.addNode(ret);
				for (Statement st : preExits)
					currentCFG.addEdge(new SequentialEdge(st, ret));

				for (VariableTableEntry entry : currentCFG.getDescriptor().getVariables())
					if (preExits.contains(entry.getScopeEnd()))
						entry.setScopeEnd(ret);
			}
		}
	}
```

I propose this fix:

```
private void addRetNodesToCurrentCFG() {
		Ret ret = new Ret(currentCFG, currentCFG.getDescriptor().getLocation());
		if (currentCFG.getNodesCount() == 0) {
			// empty method, so the ret is also the entrypoint
			currentCFG.addNode(ret, true);
		} else {
			// every non-throwing instruction that does not have a follower
			// is ending the method
			Collection<Statement> preExits = new LinkedList<>();
			for (Statement st : currentCFG.getNodes())
				if (!st.stopsExecution() && currentCFG.followersOf(st).isEmpty())
					preExits.add(st);
			if (!preExits.isEmpty()) {
				currentCFG.addNode(ret);
				for (Statement st : preExits)
					currentCFG.addEdge(new SequentialEdge(st, ret));
				for (VariableTableEntry entry : currentCFG.getDescriptor().getVariables())
					if (preExits.contains(entry.getScopeEnd()))
						entry.setScopeEnd(ret);
			}
		}
	}
```

With the fix pyLiSA runs successfully:
<img width="1384" alt="Screenshot 2023-09-21 at 15 44 39" src="https://github.com/lisa-analyzer/pylisa/assets/11473156/45c7032e-b4c0-4fc7-a3af-1c13c7bfdc06">

--- a/pylisa/src/main/java/it/unive/pylisa/PyFrontend.java
+++ b/pylisa/src/main/java/it/unive/pylisa/PyFrontend.java
@@ -480,22 +480,21 @@ public class PyFrontend extends Python3ParserBaseVisitor<Object> {
 	}
 
 	private void addRetNodesToCurrentCFG() {
-		if (currentCFG.getAllExitpoints().isEmpty()) {
-			Ret ret = new Ret(currentCFG, currentCFG.getDescriptor().getLocation());
-			if (currentCFG.getNodesCount() == 0) {
-				// empty method, so the ret is also the entrypoint
-				currentCFG.addNode(ret, true);
-			} else {
-				// every non-throwing instruction that does not have a follower
-				// is ending the method
-				Collection<Statement> preExits = new LinkedList<>();
-				for (Statement st : currentCFG.getNodes())
-					if (!st.stopsExecution() && currentCFG.followersOf(st).isEmpty())
-						preExits.add(st);
+		Ret ret = new Ret(currentCFG, currentCFG.getDescriptor().getLocation());
+		if (currentCFG.getNodesCount() == 0) {
+			// empty method, so the ret is also the entrypoint
+			currentCFG.addNode(ret, true);
+		} else {
+			// every non-throwing instruction that does not have a follower
+			// is ending the method
+			Collection<Statement> preExits = new LinkedList<>();
+			for (Statement st : currentCFG.getNodes())
+				if (!st.stopsExecution() && currentCFG.followersOf(st).isEmpty())
+					preExits.add(st);
+			if (!preExits.isEmpty()) {
 				currentCFG.addNode(ret);
 				for (Statement st : preExits)
 					currentCFG.addEdge(new SequentialEdge(st, ret));
-
 				for (VariableTableEntry entry : currentCFG.getDescriptor().getVariables())
 					if (preExits.contains(entry.getScopeEnd()))
 						entry.setScopeEnd(ret);
@@ -903,6 +902,8 @@ public class PyFrontend extends Python3ParserBaseVisitor<Object> {
 			return this.visitIf_stmt(ctx.if_stmt());
 		else if (ctx.classdef() != null)
 			return this.visitClassdef(ctx.classdef());
+		else if (ctx.decorated() != null)
+			return this.visitDecorated(ctx.decorated());
 		else
 			throw new UnsupportedStatementException("Statement " + ctx + " not yet supported");
 	}

--- a/pylisa/src/main/java/it/unive/pylisa/PyFrontend.java
+++ b/pylisa/src/main/java/it/unive/pylisa/PyFrontend.java
@@ -902,8 +902,6 @@ public class PyFrontend extends Python3ParserBaseVisitor<Object> {
 			return this.visitIf_stmt(ctx.if_stmt());
 		else if (ctx.classdef() != null)
 			return this.visitClassdef(ctx.classdef());
-		else if (ctx.decorated() != null)
-			return this.visitDecorated(ctx.decorated());
 		else
 			throw new UnsupportedStatementException("Statement " + ctx + " not yet supported");
 	}

